src/java.base/share/classes/java/io/Reader.java

```diff
@@ -140,6 +140,127 @@ public void close() {
+     * {@code CharSequence}.
+     *
+     * Reading starts at the first character in the sequence.
+     *
+     * <p> The resulting reader is not safe for use by multiple
+     * concurrent threads. If the reader is to be used by more than one
+     * thread it should be controlled by appropriate synchronization.
+     *
+     * <p> If the sequence changes while the reader is open, e.g. the length
+     * changes, the behavior is undefined.
+     *
+     * <p> The returned reader is initially open. The reader is closed by
+     * calling the {@code close()} method. Subsequent calls to {@code close()}
+     * have no effect. After the reader has been closed, the {@code read()},
+     * {@code read(char[])}, {@code read(char[], int, int)},
+     * {@code read(CharBuffer)}, {@code ready()}, {@code skip(long)}, and
+     * {@code transferTo()} methods all throw {@code IOException}.
+     *
+     * <p> The returned reader supports the {@link #mark mark()} and
+     * {@link #reset reset()} operations.
+     *
+     * @param cs {@code CharSequence} providing the character stream.
+     * @return a {@code Reader} which reads characters from {@code cs}
+     * @throws NullPointerException if {@code cs} is {@code null}
+     *
+     * @since 24
+     */
+    public static Reader of(final CharSequence cs) {
+        Objects.requireNonNull(cs);
+
+        return new Reader() {
+            private boolean isClosed;
+            private int next = 0;
+            private int mark = 0;
+
+            /** Check to make sure that the stream has not been closed */
+            private void ensureOpen() throws IOException {
+                if (isClosed)
+                    throw new IOException("Stream closed");
+            }
+
+            @Override
+            public int read() throws IOException {
+                ensureOpen();
+                if (next >= cs.length())
+                    return -1;
+                return cs.charAt(next++);
+            }
+
+            @Override
+            public int read(char[] cbuf, int off, int len) throws IOException {
+                ensureOpen();
+                Objects.checkFromIndexSize(off, len, cbuf.length);
+                if (len == 0) {
+                    return 0;
+                }
+                int length = cs.length();
+                if (next >= length)
+                    return -1;
+                int n = Math.min(length - next, len);
+                switch (cs) {
+                    case String s -> s.getChars(next, next + n, cbuf, off);
+                    case StringBuilder sb -> sb.getChars(next, next + n, cbuf, off);
+                    case StringBuffer sb -> sb.getChars(next, next + n, cbuf, off);
+                    case CharBuffer cb -> cb.get(next, cbuf, off, n);
+                    default -> {
+                        for (int i = next, j = next + n; i < j;)
+                            cbuf[off++] = cs.charAt(i++);
```

Comment:
The single increment is fine and readable. 
Introducing a local variable for the end (j) seems unnecessary and its name is not indicative of its purpose as a limit.

Reply:
I would say this current version is fine: that's how we handle encode/decode loop in classes like `String`.

Reply:
I agree with Roger - the previous version you had:

```java
for (int i = 0; i < n; i++)
    cbuf[off + i] = cs.charAt(next + i);

```
was trivial, easy to understand immediately without having to stare at it for a while to see what's being done there, compared to the current version:

```java
for (int i = next, j = next + n; i < j;)
    cbuf[off++] = cs.charAt(i++);
```
I prefer switching to the former version.

Reply:
switched to previous version to end discussion